pkg/webui/components/qr/index.js

```diff
@@ -0,0 +1,190 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/* eslint-disable no-invalid-this */
+
+import React, { useCallback } from 'react'
+import * as jsQR from 'jsqr'
+import { defineMessages } from 'react-intl'
+
+import ErrorMessage from '@ttn-lw/lib/components/error-message'
+import Message from '@ttn-lw/lib/components/message'
+
+import PropTypes from '@ttn-lw/lib/prop-types'
+import { configSelector } from '@ttn-lw/lib/selectors/env'
+import sharedMessages from '@ttn-lw/lib/shared-messages'
+
+import Spinner from '../spinner'
+import Button from '../button'
+
+import Video from './input/video'
+import Capture from './input/capture'
+
+import style from './qr.styl'
+
+const m = defineMessages({
+  permissionDeniedError: 'Permission Denied: Please allow access to your camera or upload a photo',
+  fetchingPermission: 'Please set camera permissions',
+})
+
+const RequirePermission = props => {
+  const { children, onRead } = props
+  const [allow, setAllow] = React.useState(true)
+
+  const handleRead = React.useCallback(() => {
+    onRead()
+  }, [onRead])
+
+  const setPermission = state => {
+    switch (state) {
+      case 'granted':
+      case 'prompt':
+        setAllow(true)
+        break
+      case 'denied':
+        setAllow(false)
+        break
+    }
+  }
+
+  const getPermissions = React.useCallback(async () => {
```

Comment:
```suggestion
  const getPermissions = useCallback(async () => {
```
Please use `useCallback` directly.